### ISSUE TITLE
[RFC]: Rollout Artifact Transfer Connector for Post-Training Workloads

### ISSUE THREAD

**Issue by huangyibo** [vllm-project/vllm#45221]
## Background

vLLM is increasingly used as the rollout engine for post-training frameworks such as slime, verl, and OpenRLHF.

During rollout generation, GPU workers produce intermediate artifacts:

- token_ids
- logprobs
- finish_reason
- response_mask

These artifacts are later consumed by post-training frameworks to construct trajectories and perform reward computation, advantage estimation, and policy updates.

---

## Motivation

Today, rollout artifacts traverse the normal vLLM output path before reaching middleware.

Large rollout artifacts must flow through:

```text
GPU Worker
    --> EngineCore
    --> API Server
    --> Transfer Queue
```

before they can be consumed by the post-training framework.

### Current Architecture (SOTA)

1. The end-to-end workflow of the vLLM rollout engine interacting with external post-training frameworks (vime/slime) is: 
2. The post-training framework, such as slime or vime, submits a prompt request to the vLLM API Server. 
3. The API Server forwards the request to EngineCore and Scheduler. 
4. The Scheduler dispatches the request to the GPU Worker and ModelRunner. 
5. The sampler generates rollout outputs such as token IDs, logprobs, and finish reasons. 
6. These outputs flow back through GPU Worker → EngineCore/Scheduler → API Server as normal RequestOutput data. 
7. After vLLM internal processing completes, the API Server publishes the artifact to the Transfer Queue. 
8. The post-training framework consumes the artifact from Transfer Queue for reward computation, advantage estimation, and training updates. The key limitation is that the artifact payload still traverses vLLM’s control path before reaching external middleware.

<img width="4516" height="2246" alt="Image" src="https://github.com/user-attachments/assets/f1e1a68c-7956-4cf1-86f1-31255de30c22" />

Characteristics:

- Artifact generation occurs on GPU workers.
- Artifact transfer happens after API server processing.
- RequestOutput carries rollout payloads.

---

### Functional Goals

- Enable GPU workers to directly export rollout artifacts.
- Support pluggable transfer backends.
- Decouple artifact transfer from RequestOutput generation.
- Preserve existing serving APIs for compatibility.
- Support asynchronous artifact consumption.

### Non-Goals

- Reward computation inside vLLM.
- PPO/GRPO/DPO implementation.
- Trainer orchestration.

---

## Proposal

<img width="4890" height="2415" alt="Image" src="https://github.com/user-attachments/assets/750d443d-2805-4353-81cc-9a5f1994d5d0" />

Introduce an Artifact Transfer Connector that allows GPU workers to directly export artifacts to external middleware. Basically, the workflow is as shown :

1. The post-training framework submits the prompt request through the vLLM API Server as before. 
2. EngineCore/Scheduler schedules the request and passes artifact-transfer metadata to the GPU Worker. 
3. The GPU Worker creates or binds a request-scoped ArtifactConnector. 
4. During decoding, the GPU Worker and sampler generate rollout outputs such as token IDs, logprobs, finish reasons, and related metadata. 
5. Instead of sending the full artifact back through EngineCore and API Server, the GPU Worker exports it directly through the ArtifactConnector to Transfer Queue. 
6. After the artifact is stored or completed, Transfer Queue returns a lightweight ArtifactHandle. 
7. The ArtifactHandle is propagated back through ArtifactConnector → EngineCore/Scheduler → API Server → slime/vime. 
8. The post-training framework uses the handle to fetch the artifact from Transfer Queue. 

**Summary**. This keeps large artifact payloads off the vLLM control path while preserving the existing request lifecycle and API semantics.

### **Key abstractions proposed**

- **Artifact**: The data-plane payload produced by vLLM rollout execution, such as token IDs, logprobs, finish reason, and masks, that will be consumed by the post-training framework.
- **ArtifactMetadata**: The scheduler-generated control metadata that tells the GPU worker what artifact to export, where to export it, and under which transfer policy.
- **ArtifactConnector**: The vLLM worker-side integration layer that collects rollout outputs and exports them to the configured transfer backend.
- **ArtifactBackend**: The pluggable transport or storage implementation, such as Transfer Queue, Mooncake, shared memory, or a dummy backend, that stores and serves artifacts.
- **ArtifactHandle**: The lightweight control-plane reference returned to slime/vime so the framework can locate and fetch the exported artifact from the backend.
- **ArtifactTransferConfig**: The user/framework-facing configuration that enables artifact transfer and specifies backend type, transfer mode, and exported tensor fields.

**Design principle:**

```text
Large Payloads
    --> Data Plane

ArtifactHandle
    --> Control Plane
```

---

## High-Level Abstractions

- Scheduler Side

Scheduler constructs request-scoped metadata:

```python
ArtifactMetadata(
    request_id,
    backend,
    transfer_mode,
)
```

- GPU Worker Side

```python
connector = ArtifactConnector(metadata)

connector.export(
    token_ids,
    logprobs,
)
```

- Backend Interface

```python
class ArtifactBackend:
    def put_artifact(...)
    def complete_artifact(...)
    def get_artifact(...)
```

Initial implementations:
  - TransferQueueBackend
  - MooncakeBackend
  - DummyBackend

---

## Implementation Plan

- Introduce:
```text
vllm/distributed/artifact_transfer/
```
- Add scheduler-side metadata propagation.
- Implement Transfer Queue backend.
- Add worker-side artifact export.

---

### Unit Tests

- ArtifactMetadata propagation
- ArtifactHandle generation
- Artifact lifecycle correctness

### Integration Tests

Validate:

```text
vime/slime
    <-> vLLM
    <-> Transfer Queue
```

### Performance Evaluation

Compare:

```text
Current:
GPU Worker → EngineCore → API Server → Transfer Queue

Proposed:
GPU Worker → Artifact Connector → Transfer Queue
```

Measure latency/throughput, CPU utilization, API server serialization overhead, and data copies.

---

## Expected Benefits

- Cleaner control/data plane separation.
- Reduced rollout serialization overhead.
- Extensible backend architecture.
- Alignment with KV Transfer design principles.

### Proposed Change.

This RFC proposes adding an Artifact Transfer Connector to vLLM so GPU workers can directly export rollout artifacts to external middleware (e.g., Transfer Queue, Mooncake Store) while returning only lightweight artifact handles back to the post-training frameworks through the existing vLLM control path. Reduce two-hop data movement overhead due to bypassing vLLM API server and EngineCore in the data path.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by dodatboii** [comment]
Thanks for putting together this RFC — the direction makes a lot of sense to us and aligns well with how I've been thinking about this problem.

I work in the team behind TransferQueue, which is one of the proposed ArtifactBackend implementations here. A few thoughts from my side:

**On the overall design**

The control/data plane split (`ArtifactHandle` on the control plane, actual payload on the data plane) exactly fits the design intent of TQ. Right now the common pattern in verl/vime/slime is to have vLLM push the full output through RequestOutput → API server → TQ client, which means large tensors are unnecessarily serialized and copied through layers that were designed for lightweight control messages. Moving the export point down to the GPU worker removes that overhead entirely.

**On the ArtifactBackend interface**

The three-method interface (put_artifact, complete_artifact, get_artifact) maps cleanly onto TQ's existing semantics. TQ already separates the "write metadata + allocate index" step from the "write data" step, so complete_artifact being distinct from put_artifact is a natural fit.
When implementing TransferQueueBackend, I'd suggest using `BatchMeta` as the content of `ArtifactHandle`. It already carries the fields needed to locate and retrieve data (partition_ids + global_indexes), is picklable, and integrates directly with `async_get_data()` on the consumer side.


-----
Following this RFC with great interest and looking forward to seeing this take shape~

**Comment by 0oshowero0** [comment]
Thanks for the great work! This abstraction can be helpful for ensuring that the inference engine and RL framework run smoothly.

Some comments:
1. We may need to let the Scheduler notifies what artifacts should be put into the connector, maybe add some new fields in `ArtifactMetadata`?
2. For layerwise artifact outputs (such as routed expert index), we may hope to concurrently transfer the data while computing for the next layer.
3. For worker side integration, I'm thinking about we can refer to the `KVConnector` implementation that use a simple wrapper to perform data transport:
https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/attention/kv_transfer_utils.py#L15

**Comment by stmatengss** [comment]
Nice RFC. A few thoughts from the Mooncake side:

1. ArtifactMetadata is a bit thin. It should also carry the tensor schema (which fields to export, dtype, shape, layerwise vs final), otherwise, the backend has to guess or every connector ends up hardcoding the slime/verl contract. +1 to @0oshowero0's point.
2. Layerwise streaming. For things like routed expert indices or per-layer logits we'd want to overlap export with the next layer's compute. The current connector.export(token_ids, logprobs) shape is a one-shot call — worth leaving room for a streaming variant (e.g. begin/append/complete) before locking the API.
3. TP/PP ownership. Who actually issues put_artifact under TP>1 or PP? rank0 gather + single export is simpler but wastes the parallel bandwidth; per-rank export needs the backend to understand grouping. Would be good to spell this out in the RFC. For Mooncake we'd lean on the group semantics we already added in [mooncake#2180](https://github.com/kvcache-ai/Mooncake/pull/2180).
4. Mooncake backend. Happy to contribute a MooncakeBackend implementation once the interface stabilizes — the put/complete/get split maps cleanly onto Mooncake Store's object API, and ArtifactHandle can just wrap (object_key, replica_list).

Overall +1, the data/control plane split is the right call.

CC. @ykwd @zxpdemonio

**Comment by aoshen02** [comment]
## Concrete R3 implementation and follow-up roadmap

Related implementation: vllm-project/vllm#45635

The following narrows this RFC into a concrete first implementation and staged follow-ups. R3 with a local SHM backend is the first artifact; Mooncake, joint readiness, and additional fields remain separate follow-up work.

# Artifact Connector: block-keyed storage for routed-expert outputs

## Summary

This PR introduces the first **Artifact Connector** implementation for vLLM.
The first artifact is routed-expert output (R3): the expert IDs selected for
every token executed by an MoE model.

R3 cannot remain a request-local GPU buffer when prefix caching and async
scheduling are enabled:

- a prefix-cache hit skips model execution, so cached tokens produce no new R3;
- physical GPU/KV slots are reused and are not stable object identities;
- speculative decoding can execute rows that are later rejected;
- one result may contain cached blocks, new blocks, and a partial tail.

The connector stores immutable R3 blocks using KV-compatible block hashes. KV
and R3 for the same prefix can therefore be reused together. This PR provides a
bounded local SHM backend; distributed storage and more artifact types are
follow-ups.

## Why an Artifact Connector?

The model capture hook only obtains a tensor from the current forward pass. A
separate lifecycle layer must decide:

- which logical tokens are accepted;
- which blocks are reusable across requests;
- when temporary GPU/CPU buffers can be released;
- where immutable blocks are stored; and
- how cached and new blocks form one ordered result.

These responsibilities also apply to future DSA and logprobs artifacts. The
connector keeps them out of the model runner and gives storage backends one
common object lifecycle. It is not a second KV Connector or a generic Scheduler
sidecar.

## Architecture

```text
+-------------------------------------------------------------------+
| API / OutputProcessor                                             |
| Current: return complete R3                                       |
| Follow-up: return ordered artifact keys                           |
+-------------------------------+-----------------------------------+
                                |
                                v
+-------------------------------------------------------------------+
| EngineCore                                                        |
| Scheduler -> ArtifactSchedulerConnector                           |
|              - logical request progress                           |
|              - KV-compatible hashes                              |
|              - terminal events                                   |
+-------------------------------+-----------------------------------+
                                | per-step metadata
                                v
+-------------------------------------------------------------------+
| GPU Worker                                                        |
| Model Runner -> router capture -> async GPU-to-CPU copy            |
|              -> ArtifactWorkerConnector                           |
|                 - accepted-row assembly                           |
|                 - block publication/materialization               |
+-------------------------------+-----------------------------------+
                                |
                  +-------------+-------------+
                  |                           |
                  v                           v
     +--------------------------+  +--------------------------+
     | Current                  |  | Follow-up                |
     | Bounded local SHM        |  | Mooncake distributed     |
     | background publication   |  | immutable objects        |
     +--------------------------+  +--------------------------+
```

The Scheduler connector owns logical progress and hashes, not payloads. The
Worker connector owns accepted R3 rows and publication. The Store owns opaque
immutable objects and retention, but knows nothing about tokens or speculative
decoding.

## Artifact types

An execution artifact is an auxiliary model output that is aligned to a stable
logical coordinate and may need to outlive the worker step that produced it.

| Artifact | Coordinate | Reuse | Status |
| --- | --- | --- | --- |
| Routed experts (R3) | Executed token | Prefix block | This PR |
| DSA routing/index data | Executed token | Prefix block | Follow-up |
| Logprobs/prompt logprobs | Predicted token | Prefix block | Follow-up |
| Multimodal auxiliary data | Request/object | Request only | Exploratory |
| Sampling candidate IDs | Sampling step | Request only | Exploratory |

Prefix-block artifacts are content-addressed and reusable with a matching
prefix. Request-only artifacts use the same storage contract but do not
participate in prefix hits. Reward values, tool results, finish reasons, and
Scheduler status remain normal request/control-plane metadata.

## Current PR scope

This PR implements:

- Model Runner V2 R3 capture;
- block-keyed SHM publication and prefix materialization;
- async scheduling, chunked prefill, speculative rejection, and preemption;
- background publication, bounded capacity, and KV-aligned retention;
- terminal-only R3 delivery for streaming output;
- fail-closed checks for unsupported configurations.

It does not implement Mooncake, public artifact keys, additional artifact
fields, PD disaggregation, or new parallel topologies.

When disabled, no artifact capture buffer, Store, per-step metadata, or R3 copy
is created on the normal serving path.

## Follow-ups

```text
This PR
  R3 lifecycle + local SHM
        |
        v
Unified object/profile contract + ordered keys
        |
        v
Direct Mooncake backend
        |
        v
Joint KV + mandatory-artifact prefix readiness
        |
        v
DSA, logprobs, and additional topology support
```

The main planned work is:

1. define a backend-independent object/profile contract and ordered key output;
2. add a direct Mooncake backend for external trainer/JAX consumers;
3. admit remote prefix hits only when KV and mandatory artifacts are ready;
4. add DSA and logprobs through small field adapters, not new lifecycle code;
5. validate CPU KV offload, PD, PP/DCP/PCP, and multi-writer deployments before
   enabling them.

## Validation

Development validation covers focused unit tests, async scheduling, forced
preemption, chunked prefill, speculative rejection, cold/hot prefix reuse,
streaming aggregation, SHM capacity/LRU behavior, pre-commit, and GB200/H200
throughput experiments reported in the PR discussion.

The latest update also moves R3 GPU-to-CPU transfer onto the existing async
output copy stream, avoiding a synchronous copy on the model execution critical
path.

AI assistance was used for implementation, testing, performance experiments,
and review. The submitting human reviewed the accepted changes and validation.
